File: src/main/java/com/dh/ondot/schedule/domain/service/ScheduleQueryService.java
package com.dh.ondot.schedule.domain.service;

import com.dh.ondot.core.util.DateTimeUtils;
import com.dh.ondot.schedule.core.exception.NotFoundScheduleException;
import com.dh.ondot.schedule.domain.Schedule;
import com.dh.ondot.schedule.domain.repository.ScheduleRepository;
import com.dh.ondot.schedule.infra.ScheduleQueryRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.stereotype.Service;

import java.time.Instant;

@Service
@RequiredArgsConstructor
public class ScheduleQueryService {
    private final ScheduleRepository scheduleRepository;
    private final ScheduleQueryRepository scheduleQueryRepository;

    public Schedule findScheduleById(Long id) {
        return scheduleRepository.findById(id)
                .orElseThrow(() -> new NotFoundScheduleException(id));
    }

    public Schedule findScheduleByIdEager(Long scheduleId) {
        return scheduleQueryRepository.findScheduleById(scheduleId)
                .orElseThrow(() -> new NotFoundScheduleException(scheduleId));
    }

    public Schedule findScheduleByMemberIdAndId(Long memberId, Long scheduleId) {
        return scheduleQueryRepository.findScheduleByMemberIdAndId(memberId, scheduleId)
                .orElseThrow(() -> new NotFoundScheduleException(scheduleId));
    }

    public Slice<Schedule> getActiveSchedules(Long memberId, Pageable page) {
        Instant now = DateTimeUtils.nowSeoulInstant();
        return scheduleQueryRepository.findActiveSchedulesByMember(memberId, now, page);
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/domain/service/ScheduleQueryService.kt
package com.dh.ondot.schedule.domain.service

import com.dh.ondot.core.util.TimeUtils
import com.dh.ondot.schedule.core.exception.NotFoundScheduleException
import com.dh.ondot.schedule.domain.Schedule
import com.dh.ondot.schedule.domain.repository.ScheduleRepository
import com.dh.ondot.schedule.infra.ScheduleQueryRepository
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Slice
import org.springframework.stereotype.Service

@Service
class ScheduleQueryService(
    private val scheduleRepository: ScheduleRepository,
    private val scheduleQueryRepository: ScheduleQueryRepository,
) {
    fun findScheduleById(id: Long): Schedule =
        scheduleRepository.findById(id)
            .orElseThrow { NotFoundScheduleException(id) }

    fun findScheduleByIdEager(scheduleId: Long): Schedule =
        scheduleQueryRepository.findScheduleById(scheduleId)
            .orElseThrow { NotFoundScheduleException(scheduleId) }

    fun findScheduleByMemberIdAndId(memberId: Long, scheduleId: Long): Schedule =
        scheduleQueryRepository.findScheduleByMemberIdAndId(memberId, scheduleId)
            .orElseThrow { NotFoundScheduleException(scheduleId) }

    fun getActiveSchedules(memberId: Long, page: Pageable): Slice<Schedule> {
        val now = TimeUtils.nowSeoulInstant()
        return scheduleQueryRepository.findActiveSchedulesByMember(memberId, now, page)
    }
}
